### PR TITLE
refactor /.building by using git submodules

### PR DESCRIPTION
**Pull request by aaptel**
- remove all the LuaJIT-xyz dirs
- remove .building/ 7zip archive
- put a LuaJIT git submodule in /.building
- put a lua-enet git submodule
- rewrite lua-enet cmake file
- NEED TO ACCEPT PR https://github.com/Ismoh/lua-enet/pull/1 first!!
- replace VS solution for lua_noitamp_native with cmake
- add a .building/build.py script to rebuild lua and all the lua modules, copy them in mods and optionnaly in Noita game dir

To use noita-mp after this:
- merge the cmake fix in lua-enet repo https://github.com/Ismoh/lua-enet/pull/1
- pull it here
- cd .building && python build.py --lua-version v2.0

LuaJIT v2.0 branch doesn't trigger the AddScriptColor() bug and doesn't seem to leak memory (that I could see, so far...) so that is the one we should use now.

I've also opened an issue on LuaJIT repo, lets see where it goes... https://github.com/LuaJIT/LuaJIT/issues/1136

### CODE REVIEW

**Comment by Ismoh** [comment]
PR on lua-enet was merged. I'll merge this PR now. Or any objections?

File: .building/build.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+import io
+import time
+import sys
+
+def find_git_root():
+    d = os.path.dirname(__file__)
+    while d:
+        if os.path.exists(d + r'\.git'):
+            return d
+        d = os.path.realpath(d + r'\..')
```
Comment (Furrior) [inline .building/build.py]: Can stuck in an infinite loop if the repo is downloaded via zip, so doesnt gave a .git folder cuz `C:/..` gives no runtimes and returns `C:/`

File: .building/build.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+LUA_DLL = os.path.join(SCRIPT_DIR, "LuaJIT", "src", "lua51.dll")
+LUA_LIB = os.path.join(SCRIPT_DIR, "LuaJIT", "src", "lua51.lib")
+
+def main():
+    ap = argparse.ArgumentParser("lua dependency rebuilder. By default rebuild everything with the current state of LuaJIT dir.")
+    ap.add_argument("--lua-version", help="LuaJIT version to checkout and rebuild with (git rev or branch)")
+    ap.add_argument("--build", choices=["all", "lua", "enet", "native"], help="Only rebuild one project (default all)", default="all")
+    ap.add_argument("--install-lua", action="store_true", help="Copy lua dll to noita game directory (default False)", default=False)
+    ap.add_argument("--noita-dir", help="Path to noita game dir (default %s)" % NOITA_DIR, default=NOITA_DIR)
+    args = ap.parse_args()
```
Comment (Furrior) [inline .building/build.py]: Non typed arguments 🤓

File: .building/build.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        lua_noitamp_native = build_lua_noitamp_native()
+        if not lua_noitamp_native:
+            err("LUA NOITAMP NATIVE BUILD FAILED!")
+            exit(1)
+
+    msg("ALL BUILD OK!")
+    if lua_enet:
+        copy_file(lua_enet,           os.path.join(GIT_DIR, "mods", "noita-mp", "lua_modules", "lib", "lua", "5.1"))
+    if lua_noitamp_native:
+        copy_file(lua_noitamp_native, os.path.join(GIT_DIR, "mods", "noita-mp", "lua_modules", "lib", "lua", "5.1"))
```
Comment (Furrior) [inline .building/build.py]: Why not just move `os.path.join(GIT_DIR, "mods", "noita-mp", "lua_modules", "lib", "lua", "5.1"))` to a variable?

File: .building/build.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    if args.install_lua and lua_dll:
+        copy_file(lua_dll, args.noita_dir)
+
+def build_lua(version=None, debug=True):
+    git_cmd = ""
+    if version is not None:
+        git_cmd = "&& git checkout %s"%version
```
Comment (Furrior) [inline .building/build.py]: F string not only look better, but also are **a bit** faster than % formatting 🤓 
Especially considering that you used them later

File: .building/build.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+        r'msbuild lua_noitamp_native.sln'
+    ]
+
+    build_dir = os.path.join(SCRIPT_DIR, "lua_noitamp_native", "build")
+    rmtree(os.path.join(SCRIPT_DIR, 'lua_noitamp_native', 'build'))
+    print('\n'.join(cmds))
+    os.system(' && '.join(cmds))
+    return find_dll(build_dir, "lua_noitamp_native.dll")
+
+def rmtree(path):
```
Comment (Furrior) [inline .building/build.py]:
```suggestion
def rmtree(path: str):
```

File: .building/build.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    if os.path.exists(path):
+        if os.path.isdir(path):
+            shutil.rmtree(path)
+        else:
+            os.unlink(path)
+
+def copy_file(src, dst):
```
Comment (Furrior) [inline .building/build.py]:
```suggestion
def copy_file(src: str, dst: str):
```

File: .building/build.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    print("COPY", os.path.basename(src), "->", dst)
+    if os.path.exists(dst) and not os.path.isdir(dst):
+        rmtree(dst)
+    shutil.copyfile(src, dst)
+
+def find_dll(dir, name):
+    for root, dirs, names in os.walk(dir):
+        if name in names:
+            return os.path.join(root, name)
+    return None
```
Comment (Furrior) [inline .building/build.py]:
```suggestion
def find_dll(dir: str, name: str) -> str | None:
    for root, dirs, names in os.walk(dir):
        if name in names:
            return os.path.join(root, name)
    return None
```

File: .building/build.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+
+def msg(s):
+    print("\x1b[32m%s\x1b[0m" % s)
+
+def err(s):
+    print("\x1b[31m%s\x1b[0m" % s)
```
Comment (Furrior) [inline .building/build.py]:
```suggestion
def msg(s: str):
    print(rf"\x1b[32m{s}\x1b[0m")

def err(s: str):
    print(rf"\x1b[31m{s}\x1b[0m")
```

File: .building/build.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+        r'msbuild lua-enet.sln'
+    ]
+
+    build_dir = os.path.join(SCRIPT_DIR, "lua-enet", "build")
+    rmtree(build_dir)
+    print('\n'.join(cmds))
+    os.system(' && '.join(cmds))
+    return find_dll(build_dir, "enet.dll")
+
+def build_lua_noitamp_native():
```
Comment (Furrior) [inline .building/build.py]:
```suggestion
def build_lua_noitamp_native() -> str | None:
```

File: .building/build.py [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+    build_dir = os.path.join(SCRIPT_DIR, "LuaJIT")
+    rmtree(LUA_DLL)
+    rmtree(LUA_LIB)
+    cmd = f'""{VSDEVCMD_PATH}"" && cd LuaJIT {git_cmd} && cd src && msvcbuild.bat {debug}'
+    print(cmd)
+    os.system(cmd)
+
+    return find_dll(build_dir, "lua51.dll")
+
+def build_lua_enet():
```
Comment (Furrior) [inline .building/build.py]:
```suggestion
def build_lua_enet() -> str | None:
```

File: .building/build.py (hunk 8 above)
Comment (aaptel) [reply .building/build.py]: cannot use 'r' here

Comment (Furrior) [reply .building/build.py]: Ah, its for color. Then r isnt needed, yeah

**Comment by Ismoh** [comment]
Thank you so much for all your work! @aaptel

**Pull request merged**